### ISSUE TITLE
TFR incompatible with BatchNorm layer updates

### ISSUE THREAD

**Issue by gabegrand** [tensorflow/ranking#97]
I have a custom `tf.keras.Model` that I want to train on a ranking task. I've managed to get the model integrated with TFR by calling the model inside `score_fn`. However, the model contains Keras BatchNorm layers, which require that the `update_ops` for the `moving_mean` and `moving_variance` variables are manually grouped with `train_op` (see https://www.tensorflow.org/api_docs/python/tf/layers/BatchNormalization).

Without these `update_ops`, the BN layers are not updated during training, which destroys the accuracy of the network. This is a known issue (see https://github.com/tensorflow/tensorflow/issues/19643).

When I use the model inside a vanilla Estimator (without TFR), I'm able to do something like this:

```
def model_fn(features, labels, mode, params):

    # CustomKerasModel subclasses tf.keras.Model and contains tf.keras.layers.BatchNormalization layers
    net = CustomKerasModel(params)
    logits = net(features, training=(mode == tf.estimator.ModeKeys.TRAIN))
    loss_op = mean_squared_error(labels, logits)
    
    # Necessary to manually get update_ops to update BN layers
    update_ops = net.get_updates_for(inputs)
    with tf.control_dependencies([update_ops])
        train_op = optimizer.minimize(loss_op, global_step=tf.train.get_global_step())

    return tf.estimator.EstimatorSpec(
            mode=mode,
            predictions=logits,
            loss=loss_op,
            train_op=train_op)
```

However, with TFR, the `train_op` is isolated from the `score_fn` such that there's no obvious way to use the model's `update_ops` as `control_dependencies`.

```
def make_score_fn():
    def _score_fn(context_features, group_features, mode, params, config):
        net = CustomKerasModel(params)
        logits = net(inputs, training=(mode == tf.estimator.ModeKeys.TRAIN))

        # `update_ops` are isolated from `train_op`
        update_ops = net.get_updates_for(inputs)
        return logits
    return _score_fn


def model_fn_ranking(features, labels, mode, params):

    def _train_op_fn(loss):
        """Defines train op used in ranking head."""
        return tf.contrib.layers.optimize_loss(
            loss=loss,
            global_step=tf.train.get_global_step(),
            learning_rate=params.learning_rate,
            optimizer="Adam")

    ranking_head = tfr.head.create_ranking_head(
        loss_fn=tfr.losses.make_loss_fn("pairwise_logistic_loss"),
        eval_metric_fns=eval_metric_fns(),
        train_op_fn=_train_op_fn)

    return tfr.model.make_groupwise_ranking_fn(
        group_score_fn=make_score_fn(), 
        group_size=1,
        transform_fn=None,
        ranking_head=ranking_head)
```

In https://github.com/tensorflow/ranking/issues/30, it's suggested that there's some way to add `tf.control_dependencies` for BatchNorm and Dropout layers. Can someone please clarify how I might accomplish this?

**Comment by ramakumar1729** [comment]
@gabegrand , TF v1 ops allowed for something like this. While this works in TF2.0 with backward compatibility, I'm not sure if native Keras models link to the op `tf.compat.v1.GraphKeys.UPDATE_OPS`.

```
 def _train_op_fn(loss):
    """Defines train op used in ranking head."""
    update_ops = tf.compat.v1.get_collection(tf.compat.v1.GraphKeys.UPDATE_OPS)
    minimize_op = optimizer.minimize(
        loss=loss, global_step=tf.compat.v1.train.get_global_step())
    train_op = tf.group([minimize_op, update_ops])
    return train_op
```

My recommendation would be to have your custom keras model transformations _before_ calling the scoring function, or the groupwise ranking model builder. We do provide the transform_fn closure, which is one of the arguments for `make_groupwise_ranking_fn`, where you can define this transformation. Note that this logic applies on the entire list, not just the groups defined in the score function.

**Comment by ramakumar1729** [comment]
Linking to #32 for reference.

**Comment by gabegrand** [comment]
@ramakumar1729 Thanks so much for the suggestion! You're correct that native Keras models do not add to `tf.compat.v1.GraphKeys.UPDATE_OPS`, so I had to add the variables to the collection manually. Once I did so, I was able to get the BN variables to update properly using your solution.

For others who may encounter this issue, here's the solution I used:

```
def make_score_fn():
    def _score_fn(context_features, group_features, mode, params, config):
        net = CustomKerasModel(params)
        logits = net(inputs, training=(mode == tf.estimator.ModeKeys.TRAIN))
        update_ops = net.get_updates_for(inputs)

        # Hack to add BatchNorm update ops to graph
        for op in net.get_updates_for(inputs):
            if "batch_normalization" in op.name:
                tf.add_to_collection(name=tf.compat.v1.GraphKeys.UPDATE_OPS, value=op)

        return logits
    return _score_fn


def model_fn_ranking(features, labels, mode, params):

    def _train_op_fn(loss):
        """Defines train op used in ranking head."""
        update_ops = tf.compat.v1.get_collection(tf.compat.v1.GraphKeys.UPDATE_OPS)
        minimize_op = tf.contrib.layers.optimize_loss(
            loss=loss,
            global_step=tf.train.get_global_step(),
            learning_rate=params.learning_rate,
            optimizer="Adam")
        train_op = tf.group([minimize_op, update_ops])
        return train_op

    ranking_head = tfr.head.create_ranking_head(
        loss_fn=tfr.losses.make_loss_fn("pairwise_logistic_loss"),
        eval_metric_fns=eval_metric_fns(),
        train_op_fn=_train_op_fn)

    return tfr.model.make_groupwise_ranking_fn(
        group_score_fn=make_score_fn(), 
        group_size=1,
        transform_fn=None,
        ranking_head=ranking_head)
```
